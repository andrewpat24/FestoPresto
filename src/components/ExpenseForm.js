import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? props.expense.amount / 100 : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
      expenseIsNew: this.props.expenseIsNew
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({
      note
    }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount
      }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt)
      this.setState(() => ({
        createdAt
      }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Amount and Description must be filled in."
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}

        <input
          autoFocus
          type="text"
          className="text-input"
          placeholder="Description"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder="add a note for your expense (optional)."
          className="textarea"
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          {this.state.expenseIsNew ? (
            <button className="btn md blue-bg">Add Expense</button>
          ) : (
            <button className="btn md blue-bg">Edit Expense</button>
          )}
        </div>
      </form>
    );
  }
}

ExpenseForm.defaultProps = {
  expenseIsNew: true
};

export default ExpenseForm;
