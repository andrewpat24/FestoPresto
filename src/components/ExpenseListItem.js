import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

const ExpenseListItem = props => {
  const { id, description, amount = 100, createdAt = 500 } = props;
  return (
    <Link className="list-item" to={`/edit/${id}`}>
      <div>
        <h3 className="list-item__title">{description}</h3>
        <span className="list-item__subtitle">
          {moment(createdAt).format("MMMM Do, YYYY")}
        </span>
      </div>
      <div>
        <h3 className="list-item__data">
          {numeral(amount / 100).format("$0,0.00")}
        </h3>
      </div>
    </Link>
  );
};

export default ExpenseListItem;
