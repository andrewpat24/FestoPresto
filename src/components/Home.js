import React from "react";

export const Home = () => (
  <div className="home">
    <h1>HOME PAGE</h1>
  </div>
);

export default Home;
