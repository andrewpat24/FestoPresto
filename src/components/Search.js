import React from "react";

export const Search = () => (
  <div className="search">
    <h1>SEARCH PAGE</h1>
  </div>
);

export default Search;
