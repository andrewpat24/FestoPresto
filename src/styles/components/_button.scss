.btn {
  display: inline-block;
  border: none;
  padding: $s-size;
  text-decoration: none;
  line-height: 1;
  font-weight: 100;
}

.md {
  font-size: $font-size-lg;
}

.blue-bg {
  background-color: $blue;
  color: white;
}

.blue-secondary-bg {
  background-color: $blue-secondary;
  color: white;
}

.transparent-bg {
  background: none;
  color: white;
}
