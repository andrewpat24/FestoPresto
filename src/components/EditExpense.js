import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
// Helpers
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpense extends React.Component {
  onSubmit = expense => {
    const id = this.props.expense.id;
    this.props.startEditExpense(id, expense);
    this.props.history.push("/dashboard");
  };

  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push("/dashboard");
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            expenseIsNew={false}
            onSubmit={this.onSubmit}
          />
        </div>
        <div className="content-container content-space-around-s">
          <button
            className="btn md blue-secondary-bg btn-space-around"
            onClick={this.onRemove}
          >
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => {
      return expense.id === props.match.params.id;
    })
  };
};

const mapDispatchToProps = dispatch => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: expense => dispatch(startRemoveExpense(expense))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpense);
