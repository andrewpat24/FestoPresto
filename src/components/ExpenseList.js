import React from "react";
import { connect } from "react-redux";
// Components
import ExpenseListItem from "./ExpenseListItem";
// Helpers
import selectExpenses from "../selectors/expenses";

export const ExpenseList = props => {
  let expensesProp;
  if (typeof props.expenses !== "undefined") {
    expensesProp = props.expenses;
  } else {
    expensesProp = [];
  }

  return (
    <div className="content-container">
      <div className="list-header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-desktop">Expense</div>
        <div className="show-for-desktop">Amount</div>
      </div>

      <div className="list-body">
        {expensesProp.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No expenses</span>
          </div>
        ) : (
          <div>
            {expensesProp.map(expense => {
              return <ExpenseListItem {...expense} key={expense.id} />;
            })}
          </div>
        )}
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
