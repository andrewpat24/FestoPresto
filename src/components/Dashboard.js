import React from "react";
// Components
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

const Dashboard = () => (
  <section component="Dashboard">
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </section>
);

export default Dashboard;
