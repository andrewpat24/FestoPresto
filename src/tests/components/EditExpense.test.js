import React from "react";
import { shallow } from "enzyme";
import { EditExpense } from "../../components/EditExpense";
import expenses from "../fixtures/expenses";

let startEditExpense, startRemoveExpense, history, wrapper;
const { amount, createdAt, description, id, note } = expenses[0];

beforeEach(() => {
  startEditExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpense
      startEditExpense={startEditExpense}
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense={expenses[0]}
    />
  );
});

test("Should render EditExpense component correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("Should handle editExpense", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);
  expect(history.push).toHaveBeenCalledWith("/dashboard");
  expect(startEditExpense).toHaveBeenLastCalledWith(id, {
    amount,
    createdAt,
    description,
    id,
    note
  });
});

test("Should handle removeExpense", () => {
  wrapper.find("button").prop("onClick")(expenses[0]);
  expect(history.push).toHaveBeenCalledWith("/dashboard");
  expect(startRemoveExpense).toHaveBeenLastCalledWith({
    id
  });
});
