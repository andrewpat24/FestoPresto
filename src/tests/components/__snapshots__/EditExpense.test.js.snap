// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpense component correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 50,
          "createdAt": 0,
          "description": "Gum bill",
          "id": "1",
          "note": "Test note",
        }
      }
      expenseIsNew={false}
      onSubmit={[Function]}
    />
  </div>
  <div
    className="content-container content-space-around-s"
  >
    <button
      className="btn md blue-secondary-bg btn-space-around"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
